src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetPowerShellUsageCollector.cs

```diff
@@ -0,0 +1,421 @@
+
+        // Const name for emitting when VS solution close or VS instance close.
+        public const string Name = nameof(Name);
+        public const string NuGetPowershellPrefix = "NuGetPowershellPrefix."; // Using prefix prevent accidental same name property collission from different type telemetry.
+        public const string NuGetVSSolutionClose = nameof(NuGetVSSolutionClose);
+        public const string NuGetVSInstanceClose = nameof(NuGetVSInstanceClose);
+        public const string SolutionCount = nameof(SolutionCount);
+        public const string PMCPowerShellLoadedSolutionCount = nameof(PMCPowerShellLoadedSolutionCount);
+        public const string PMUIPowerShellLoadedSolutionCount = nameof(PMUIPowerShellLoadedSolutionCount);
+
+        private int _solutionCount;
+        // _vsSolutionData hold telemetry data for current VS solution session.
+        private NuGetPowershellVSSolutionCloseEvent _vsSolutionData;
+        // _vsInstanceData hold telemetry for current VS instance session.
+        private readonly NuGetPowershellVSInstanceCloseEvent _vsInstanceData;
+        private TelemetryEvent _powerShellLoadEvent;
+        private object _lock = new object();
+
+        public NuGetPowerShellUsageCollector()
+        {
+            _vsSolutionData = new NuGetPowershellVSSolutionCloseEvent();
+            _vsInstanceData = new NuGetPowershellVSInstanceCloseEvent();
+
+            NuGetPowerShellUsage.PowerShellLoadEvent += NuGetPowerShellUsage_PMCLoadEventHandler;
+            NuGetPowerShellUsage.PowerShellCommandExecuteEvent += NuGetPowerShellUsage_PowerShellCommandExecuteEvent;
+            NuGetPowerShellUsage.InitPs1LoadEvent += NuGetPowerShellUsage_InitPs1LoadEvent;
+            NuGetPowerShellUsage.PMCWindowsEvent += NuGetPowerShellUsage_PMCWindowsEventHandler;
+            NuGetPowerShellUsage.SolutionOpenEvent += NuGetPowerShellUsage_SolutionOpenHandler;
+            NuGetPowerShellUsage.SolutionCloseEvent += NuGetPowerShellUsage_SolutionCloseHandler;
+            NuGetPowerShellUsage.VSInstanceCloseEvent += NuGetPowerShellUsage_VSInstanseCloseHandler;
+        }
+
+        private void NuGetPowerShellUsage_PMCLoadEventHandler(bool isPMC)
+        {
+            AddPowerShellLoadedData(isPMC, _vsSolutionData);
+        }
+
+        internal void AddPowerShellLoadedData(bool isPMC, NuGetPowershellVSSolutionCloseEvent vsSolutionData)
+        {
+            lock (_lock)
+            {
+                // PowerShellHost loaded first time, let's emit this to find out later how many VS instance crash after loading powershell.
+                if (!vsSolutionData._loadedFromPMC && !vsSolutionData._loadedFromPMUI)
+                {
+                    var telemetryEvent = new TelemetryEvent(NuGetPowerShellLoaded, new Dictionary<string, object>
+                        {
+                            { NuGetPowershellPrefix + LoadedFromPMC, isPMC }
+                        });
```

Comment:
Why does it need to check if it's the first time?

My understanding is that we're trying to understand if the host is started, because once it's loaded, it's never unloaded, and therefore we want to understand the the process is "paying the price" of increased memory usage. However, we know that we might actually load two powershell hosts, one for PMC, one for packages.config install.ps1. Do we need to track both? If so, you have a bug because only the first of the two will be reported (in which case, what's the point of reporting if it's the PMC host or not?). This is the risk of duplicating business logic in more than one place.

Although, if the process loads both the PMC and PM UI powershell hosts, it will load the assemblies once, there will be two powershell hosts in memory on the heap. The spec had too many iterations, so I can't remember what we're trying to measure (powershell assemblies loaded, but do we care about how many powershell hosts are on the heap?).

Even if we're only measuring that the powershell assemblies are loaded, a static bool in the class will ensure it's only being reported once. But it keeps the data being reported close to the thing it's actually measuring, so there's less risk of future changes causing a bug in the other.